### PR TITLE
fix flattening time-point specific override failures

### LINKED ISSUE #363: flatten_timepoint_specific_output_overrides fails
I'm getting the following error when using the function: 
```
Traceback (most recent call last):
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 2646, in get_loc
    return self._engine.get_loc(key)
  File "pandas/_libs/index.pyx", line 111, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 997, in pandas._libs.hashtable.Int64HashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1004, in pandas._libs.hashtable.Int64HashTable.get_item
KeyError: 4525

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/petab/core.py", line 140, in flatten_timepoint_specific_output_overrides
    df_unique_values.loc[nrow, OBSERVABLE_ID])
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexing.py", line 1761, in __getitem__
    return self._getitem_tuple(key)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexing.py", line 1271, in _getitem_tuple
    return self._getitem_lowerdim(tup)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexing.py", line 1388, in _getitem_lowerdim
    section = self._getitem_axis(key, axis=i)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexing.py", line 1964, in _getitem_axis
    return self._get_label(key, axis=axis)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexing.py", line 624, in _get_label
    return self.obj._xs(label, axis=axis)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/generic.py", line 3537, in xs
    loc = self.index.get_loc(key)
  File "/home/leonard/Anaconda/envs/general/lib/python3.8/site-packages/pandas/core/indexes/base.py", line 2648, in get_loc
    return self._engine.get_loc(self._maybe_cast_indexer(key))
  File "pandas/_libs/index.pyx", line 111, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 997, in pandas._libs.hashtable.Int64HashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1004, in pandas._libs.hashtable.Int64HashTable.get_item
KeyError: 4525
```

The problem seems to be, that some indices are missing after `df_unique_values = df.drop_duplicates()`.
Changing line 137 from `for nrow in range(len(df_unique_values.index)):` to `for nrow in df_unique_values.index:` should fix this problem.

Still, I get an empty measurement df, when preequilibration conditions are NaN, because of lines 141-142 (`petab_problem.measurement_df[PREEQUILIBRATION_CONDITION_ID] <= df_unique_values.loc[nrow, PREEQUILIBRATION_CONDITION_ID]`) .

### PR DESCRIPTION
**Pull request by yannikschaelte**
fixes #363

### CODE REVIEW

File: petab/core.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import numpy as np
 import pandas as pd
 
+import petab
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Unused import petab](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 2]
```diff
@@ -116,32 +117,50 @@ def flatten_timepoint_specific_output_overrides(
...
-               df_unique_values.loc[nrow, PREEQUILIBRATION_CONDITION_ID])
-            & (petab_problem.measurement_df[SIMULATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, SIMULATION_CONDITION_ID])
+    for irow in df_unique_values.index:
+        df = measurement_df.loc[
+            (measurement_df[OBSERVABLE_ID] ==
+             df_unique_values.loc[irow, OBSERVABLE_ID])
+            & (not has_preeq or (
+               measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
+               df_unique_values.loc[irow, PREEQUILIBRATION_CONDITION_ID]))
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (add 1 space).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 3]
```diff
@@ -116,32 +117,50 @@ def flatten_timepoint_specific_output_overrides(
...
-            & (petab_problem.measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, PREEQUILIBRATION_CONDITION_ID])
-            & (petab_problem.measurement_df[SIMULATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, SIMULATION_CONDITION_ID])
+    for irow in df_unique_values.index:
+        df = measurement_df.loc[
+            (measurement_df[OBSERVABLE_ID] ==
+             df_unique_values.loc[irow, OBSERVABLE_ID])
+            & (not has_preeq or (
+               measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (add 1 space).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 4]
```diff
@@ -116,33 +116,51 @@ def flatten_timepoint_specific_output_overrides(
...
-            & (petab_problem.measurement_df[SIMULATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, SIMULATION_CONDITION_ID])
-        ]
+    for irow in df_unique_values.index:
+        df = measurement_df.loc[
+            (measurement_df[OBSERVABLE_ID] ==
+             df_unique_values.loc[irow, OBSERVABLE_ID])
+            & (not has_preeq or (
+                    measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
+                    df_unique_values.loc[irow, PREEQUILIBRATION_CONDITION_ID]))
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (add 1 space).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 5]
```diff
@@ -156,8 +174,8 @@ def flatten_timepoint_specific_output_overrides(
                 # and unique_sc[j] in their corresponding column
                 # (full-string matches are denoted by zero)
                 idxs = (
-                    df[NOISE_PARAMETERS].str.find(cur_noise) +
-                    df[OBSERVABLE_PARAMETERS].str.find(cur_sc)
+                        df[NOISE_PARAMETERS].str.find(cur_noise) +
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (remove 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 6]
```diff
@@ -116,33 +116,51 @@ def flatten_timepoint_specific_output_overrides(
...
-               df_unique_values.loc[nrow, PREEQUILIBRATION_CONDITION_ID])
-            & (petab_problem.measurement_df[SIMULATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, SIMULATION_CONDITION_ID])
-        ]
+    for irow in df_unique_values.index:
+        df = measurement_df.loc[
+            (measurement_df[OBSERVABLE_ID] ==
+             df_unique_values.loc[irow, OBSERVABLE_ID])
+            & (not has_preeq or (
+                    measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (add 1 space).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 7]
```diff
@@ -156,8 +174,8 @@ def flatten_timepoint_specific_output_overrides(
                 # and unique_sc[j] in their corresponding column
                 # (full-string matches are denoted by zero)
                 idxs = (
-                    df[NOISE_PARAMETERS].str.find(cur_noise) +
-                    df[OBSERVABLE_PARAMETERS].str.find(cur_sc)
+                        df[NOISE_PARAMETERS].str.find(cur_noise) +
+                        df[OBSERVABLE_PARAMETERS].str.find(cur_sc)
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong hanging indentation (remove 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

File: petab/core.py [hunk 8]
```diff
@@ -116,33 +116,51 @@ def flatten_timepoint_specific_output_overrides(
...
-               df_unique_values.loc[nrow, PREEQUILIBRATION_CONDITION_ID])
-            & (petab_problem.measurement_df[SIMULATION_CONDITION_ID] <=
-               df_unique_values.loc[nrow, SIMULATION_CONDITION_ID])
-        ]
+    for irow in df_unique_values.index:
+        df = measurement_df.loc[
+            (measurement_df[OBSERVABLE_ID] ==
+             df_unique_values.loc[irow, OBSERVABLE_ID])
+            & (not has_preeq or
+                (measurement_df[PREEQUILIBRATION_CONDITION_ID] <=
```
Comment (dweindl) [inline petab/core.py]: Codacy Issue found: [Wrong continued indentation (remove 1 space).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5223309)

**Comment by LeonardSchmiester** [comment]
For the Sobotta model I get:
```
Traceback (most recent call last):
  File "/home/leonard/Anaconda/envs/benchmark_collection/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3319, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-73ba67d08c8d>", line 1, in <module>
    flatten_petab_problem = flatten_timepoint_specific_output_overrides(petab_problem)
  File "/home/leonard/Anaconda/envs/benchmark_collection/lib/python3.7/site-packages/petab/core.py", line 178, in flatten_timepoint_specific_output_overrides
    df[OBSERVABLE_PARAMETERS].str.find(cur_sc)
  File "/home/leonard/Anaconda/envs/benchmark_collection/lib/python3.7/site-packages/pandas/core/strings.py", line 1954, in wrapper
    return func(self, *args, **kwargs)
  File "/home/leonard/Anaconda/envs/benchmark_collection/lib/python3.7/site-packages/pandas/core/strings.py", line 3083, in find
    result = str_find(self._parent, sub, start=start, end=end, side="left")
  File "/home/leonard/Anaconda/envs/benchmark_collection/lib/python3.7/site-packages/pandas/core/strings.py", line 1366, in str_find
    raise TypeError(msg)
TypeError: expected a string object, not float
```
Possibly due to some empty observableParameter / noiseParameter rows.

Additionally I think the function will still scale very badly with the size of the measurement file. Why do we need the <= in lines 159 & 161? Shouldn't == do the job?

**Comment by LeonardSchmiester** [comment]
Well, it doesn't throw an error anymore. But the ouput is a nonType object which I guess is not intended!? Might be the same reason as in my comment in #363?

**Pull request merged**